Add support for forwarded Tls-Client-Cert

JasonN3 opened the pull request:
This is to allow cert auth through a reverse proxy that terminates the SSL connection. 

The changes link off of the existing processes that are used to process X-Forwarded-For header and related headers. If X-Forwarded-* headers are accepted by the existing code, the existing request will be modified so the forwarded client certificate will be treated like it was the client certificate in the request. This way all existing checks still apply. This is the same process that was already being used for the remote address. Since the header that includes the forwarded certificate differs between reverse proxies, the header is configurable using the option x_forwarded_for_client_cert_header. This approach was chosen because it requires minimal changes and takes advantage of the existing code.

 I was able to compile and test this using Traefik and the header X-Forwarded-Tls-Client-Cert.

Resolves: #12178

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/vault?pullRequest=17272)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/vault?pullRequest=17272)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/vault?pullRequest=17272).</sub>

mpalmi commented:
Greetings, @JasonN3! Thanks for your submission.

Have you had a chance to look at our [Contribution Guidelines](https://github.com/hashicorp/vault/blob/main/CONTRIBUTING.md#pull-requests)? Please make sure to review and update the PR to be in compliance with these guidelines.

I will follow up with a more thorough review of the PR, but figured I'd flag the minor things to get the ball rolling on this.

In order to kick off CI, you'll need a changelog entry. The CATEGORY seems like `improvement`, since we'll be enabling a new usecase.

Other than that, a bit more description would help provide a bit more context for a review. Particularly interested in:

> Your pull request should have a description of what it accomplishes, how it does so, and why you chose the approach you did. PRs should include unit tests that validate correctness and the existing tests must pass. Follow-up work to fix tests does not need a fresh issue filed.

Please let us know if you have any questions.

cipherboy commented:
@JasonN3 To clarify @mpalmi questions a little (if I may...) I think I read #12178 as applying to the CertAuth plugin, whereas this appears to help Agents' mutual TLS.

I think the former only needs to have the mount tuning enabled, and then [read the request headers rather than the TLS cert](https://github.com/hashicorp/vault/blob/87801ecf1f7413bd288d82016f55c84d46ea4a99/builtin/credential/cert/path_login.go#L227-L230).

It'd be great to get clarity as to use case.

The other question I have is of precedence. Sometimes you'll want the LB to do mutual TLS on both ends of the connection (mTLS to client and mTLS to Vault)... But obviously there's issues with defaulting to the header by default over the mTLS value (if it is end-client supplied, this is bad). It might also warrant a discussion of if the header-provided cert should be re-verified by Vault rather than assumed to be already validated by the LB/...

JasonN3 commented:
@mpalmi, I hadn't read that. I've updated my description and added the changelog entry. Let me know if I need to add more detail to either one or if there's anything else I missed.

@cipherboy, the intent for that issue was to allow client certificate authentication through a reverse proxy. To expand on the use case and the reason I submitted that issue initially, I have a CA server that each of my servers gets a certificate from. I want my servers to be able to query Vault to get a secret automatically using that certificate as authentication. Prior to this change, I could not do that because my Vault server is behind a LB. The LB would terminate the SSL connection and start a new connection. This would cause the client cert to be in the header X-Forwarded-Tls-Client-Cert, which wasn't being read by the cert authentication method.
Do you have any examples where Vault can handle mTLS other than in cert auth? As for it being re-verified by Vault, it is. The changes I made do not modify the code that verifies the certificate. It just moves the certificate from a header to the root of the request. The rest of the processes that verify the certificate still apply. I did test this by testing the authentication using a certificate issued by a CA that Vault was not told to trust and authentication rightfully failed.

cipherboy commented:
Ah, so I see, this does indeed do client cert now that I trace it. A test might be good in that regard ;-)

I was thinking https://www.vaultproject.io/docs/agent/autoauth/methods/cert was it, but I believe this just allows Agent to interpose and provide its own cert for cert auth, rather than using the client's (and passing that through for Vault).

cipherboy commented:
@JasonN3 I'd probably add one in the cert auth tests: 

https://github.com/hashicorp/vault/blob/main/builtin/credential/cert/backend_test.go 

There's a lot of test certificates available to use, and when you call `vault.NewTestCluster` you can add a new option to the test config which enables this header and adds it to the listener:

https://github.com/hashicorp/vault/blob/87801ecf1f7413bd288d82016f55c84d46ea4a99/vault/testing.go#L1143
https://github.com/hashicorp/vault/blob/87801ecf1f7413bd288d82016f55c84d46ea4a99/vault/testing.go#L1472

You'll have to add a header to the client and invalidate the token, e.g.:

https://github.com/hashicorp/vault/blob/87801ecf1f7413bd288d82016f55c84d46ea4a99/builtin/logical/pki/backend_test.go#L5214-L5215
https://github.com/hashicorp/vault/blob/87801ecf1f7413bd288d82016f55c84d46ea4a99/api/client.go#L970

And then re-auth with the header, probably by direct `Logical().Write(...)` would be my idea. 

My 2c. :-)

JasonN3 commented:
I ended up just copying the test certificate from those test and creating 2 tests that are purely focused on the modified code.
Test 1 (reject_bad_cert_in_header) will verify that a certificate that is not properly formatted will be rejected.
Test 2 (pass_cert) will verify that a certificate that is properly formatted will be moved in to r.TLS.PeerCertificates.

Verification that the certificate is valid was not part of this modification as that is done elsewhere using the certificates normally provided in r.TLS.PeerCertificates by the client.

cipherboy commented:
Thanks @JasonN3! I'll take another look by Friday, feel free to ping me if I haven't gotten around to it...

cipherboy commented:
\o sorry about the delay @JasonN3.

I think we're going to need to discuss this a little more internally first. We've definitely talked about the idea before and would like to include the feature, but I think we need to have a few more conversations about implementation first. In particular, I think we'd worried about header use vs connection use, and what the final deployment architecture looks like (how would you ensure clients can't directly connect to the Vault system and provide header values, would you do a separate mTLS to the LB, &c).

I'll let you know what comes of those discussions, they'll hopefully occur in the next couple of weeks.

JasonN3 commented:
I understand the need to discuss it further. If there's any concerns, feel free to post them. I might have an answer for you.

As for a client directly connecting to the Vault server and providing the header, that's not possible unless the server owner specifies XForwardedForAuthorizedAddrs to include the client source addresses or the LB is configured to forward that header when received. This is the same limitations that exist for overriding the source address using X-Forwarded-For. Using the provided change, a TLS connection between the LB and the server is still required as the client certificate auth method mandates a TLS connection. It just won't use mTLS because the LB doesn't have a client certificate. Even if you were to add in mTLS between the LB and the Vault server, it wouldn't change anything the client could do. All that mTLS connection would do is verify the LB but then you still need to verify the client certificate that was passed in the header. The LB can't use the client certificate that was passed to it by the client because the LB doesn't have the private key. 

The process it goes through:
1. The client makes a request to the LB using https and specifies the client certificate's public key
2. The LB accepts (either blindly or after verifying it against a CA cert) the client certificate and moves the certificate to a header. The client will have to have a valid private key regardless of whether the certificate is checked at this point because otherwise it won't be able to encrypt the data for the TLS connection and communication will fail. The LB *should* also be configured to override/erase the Tls-Client-Cert (or equivalent) header if it was passed in.
3. The LB makes a new connection to the Vault server using https and specifies the client certificate's public key in the header
4. The Vault server ensures the source (the LB) is in the XForwardedForAuthorizedAddrs list, makes sure the connection from the LB is using https, and validates the client certificate is valid and allowed to log in. It will then generate the token with the appropriate permissions and pass it back to the LB to be passed back to the client.

cipherboy commented on http/handler.go:
```diff
@@ -513,6 +516,35 @@ func WrapForwardedForHandler(h http.Handler, l *configutil.Listener) http.Handle
+					client_certs = append(client_certs, cert)
+				}
+			}
+			r.TLS.PeerCertificates = client_certs
```
Note that, if no TLS was used between the LB and us, `r.TLS` will be nil.

cipherboy replied:
Right, it does currently, but if we were to implement this, we wouldn't necessarily need to require it if the LB and Vault were colocated on the same machine (e.g., over localhost). This is a useful deployment scenario with Vault Enterprise's Seal Wrap FIPS compliance, wherein Vault's TLS isn't necessarily FIPS compliant, but putting a FIPS-validated LB in front of it would be useful.

JasonN3 replied:
The condition I added in 7396430e21b30ae0d62d66a002ddaf268cca6e71 should keep it from panicking. It will just return an error now.

Allowing non-secured traffic between the LB and the server (even if it is on localhost) would open it up to a MITM attack or traffic sniffing. Returning an error instead of allowing a potential insecure setup would be the preferred option in my opinion. Especially in situations where FIPS is involved.

joecorall replied:
> Allowing non-secured traffic between the LB and the server (even if it is on localhost) would open it up to a MITM attack or traffic sniffing.

Not sure this is true if traffic between the LB and server are [encrypted in transit](https://cloud.google.com/docs/security/encryption-in-transit). e.g. if Vault is running in Cloud Run behind a serverless LB.

cipherboy replied:
Same would also apply for unix: https://developer.hashicorp.com/vault/docs/configuration/listener/unix 

Having root on the box is sufficient to dump Vault's memory anyways, and so there's more interesting attack vectors than intercepting (plaintext) unix socket traffic.

cipherboy reviewed (commented):
Sorry about the delay. We were initially worried about whether or not useful validation would occur on the chain, but we've been unable to reproduce the attack we were thinking of.

I think ideally, we'd like to see the LB's credentials persisted rather than potentially overwritten here, to allow us to do LB credential pinning in the future. I think as such, it makes sense to put this directly on the request, rather than on the `req.TLS.PeerCertificates` object, to allow plugins and operators to decide which they want to use.

JasonN3 commented:
> Sorry about the delay. We were initially worried about whether or not useful validation would occur on the chain, but we've been unable to reproduce the attack we were thinking of. 

Glad to hear the attacks failed. This change shouldn't open up new forms of attacks, but it's definitely better to test it.

> Could you send an unrelated chain in addition to a cert you control, wherein the unrelated chain was a valid trusted chain by Vault, but the cert you controlled would sign the TLS handshake

The certificate for the https communication is separate from the peer certificate that is used for authentication, so they can be and should be unrelated. It would be a bad idea to have a public CA trusted to validate your authentication since you can't control what certs they will sign, but you will want your https traffic to be signed by a public CA (or at least a CA the clients already trust) to validate your server. For the tests I've done, my Vault server is signed by a public CA and authentication is signed by an internal CA.

> I think ideally, we'd like to see the LB's credentials persisted rather than potentially overwritten here, to allow us to do LB credential pinning in the future. I think as such, it makes sense to put this directly on the request, rather than on the `req.TLS.PeerCertificates` object, to allow plugins and operators to decide which they want to use.

I'm confused by this. What credentials would the LB normally pass? As far as I can find, a LB can't form a mTLS connection with a backend server. It can only pass along the mTLS connection that the client tried for form with the LB or form a new TLS connection to the backend server. Do you know of any LBs that can initiate a mTLS connection to the backend server? If not, `req.TLS.PeerCertificates` can only be used by the client and not a LB so nothing would be lost when it is updated. As for plugins or operators detecting if it was used, they can still check for the header. The header is never removed. It is just copied to where the certificate normally would be if there wasn't a LB so the rest of the cert authentication code doesn't need to be modified to support this situation and can behave the same regardless of whether a LB is used.

cipherboy commented:
> > Could you send an unrelated chain in addition to a cert you control, wherein the unrelated chain was a valid trusted chain by Vault, but the cert you controlled would sign the TLS handshake
> 
> The certificate for the https communication is separate from the peer certificate that is used for authentication, so they can be and should be unrelated. It would be a bad idea to have a public CA trusted to validate your authentication since you can't control what certs they will sign, but you will want your https traffic to be signed by a public CA (or at least a CA the clients already trust) to validate your server. For the tests I've done, my Vault server is signed by a public CA and authentication is signed by an internal CA.

Right, this was the theoretical attack: can a client of the LB send a two part chain `(valid client cert, unrelated attacker cert)`, with the LB validating the TLS con.nection is signed by the attacker cert, thus passing through to Vault, and with cert-auth succeeding but verifying and approving based on the valid client cert -- even though there was no signature by its corresponding private key. The answer is no, as TLS stacks will strip out unrelated certs from the validation (at least, from our testing, Go's and OpenSSL's will).

JasonN3 commented:
> > I'm confused by this. What credentials would the LB normally pass? As far as I can find, a LB can't form a mTLS connection with a backend server.
> 
> Why can't a LB have a separate mTLS to Vault? That's a common scenario, initiating a new mTLS from LB to a backend service with its own credential, to ensure the data is still protected and authenticated (preventing non-LB from contacting backend service directly). This is perhaps more common when the service is over the network, rather than a LB running on the same instance as the backend service. See e.g., https://docs.nginx.com/nginx/admin-guide/security-controls/securing-http-traffic-upstream/ .

I hadn't come across that situation. Normally I just see TLS between the LB and server and not mTLS. In that case, yes, it would cause some information to be lost. However, it would depend on where you plan on verifying the mTLS cert. It should probably be verified before any of the X-Forwarded headers are processed. That way there is less code run before the LB client cert would be rejected. In that case it wouldn't matter if the LB client cert is lost. Alternatively, I could prepend the client cert to the list of PeerCertificates, which is the commit I just pushed (b7a1ffec43976284b21521d4a7f2e66bd00b5358). Only the first certificate is processed for login (builtin/credential/cert/path_login.go line 230) so any additional certificates would still be available, but the current code would ignore them.

JasonN3 commented:
@cipherboy, just checking in. Any next steps you need from me?

cipherboy commented:
@JasonN3 Nope! Thank you again! :-) 

We're likely going to merge this soon, I was waiting for 1.13 to branch so that we'd have time to make the changes I've discussed above (w.r.t. using a different `PeerCertificates` location to allow distinguishing the LB's mTLS certs from the remote peer's and also allowing this feature to work with non-TLS enabled LBs).

se-chris-thach commented:
Would love this feature also for my deployment. Thank you!

hmrodrigues commented:
Hey @JasonN3 

Don't know if this is the place, but I've patched this into 1.13.2 but it fails to parse the certificate when NGINX is in front from Vault. Don't know if this is a issue with my setup or if NGINX just sends the certificate in a different way. Here is my config files.

Vault agent
```
vault {
    address = "https://vault.intra.hugorodrigues.xyz"
}

auto_auth {
    method {
        type = "cert"

        config {
            client_cert = "/etc/vault/cert.pem"
            client_key = "/etc/vault/key.pem"
        }
    }
}

cache {
}

api_proxy {
    use_auto_auth_token = "force"
}

listener "tcp" {
    address = "127.0.0.1:8202"
    tls_disable = true
}
```

Vault
```
storage "file" {
	path = "/var/lib/vault"
}

ui = true

listener "tcp" {
    tls_cert_file = "/etc/vault/vault.crt"
    tls_key_file = "/etc/vault/vault.key"
    tls_min_version = "tls13"
    x_forwarded_for_authorized_addrs = "127.0.0.1"
    x_forwarded_for_client_cert_header = "X-SSL-CERT"
}

api_addr = "https://127.0.0.1:8200"
```

NGINX server block
```
server {
    
    listen 443 ssl http2;
    server_name vault.intra.hugorodrigues.xyz;

    ssl_certificate /etc/nginx/ssl/vault.intra.hugorodrigues.xyz.crt;
    ssl_certificate_key /etc/nginx/ssl/vault.intra.hugorodrigues.xyz.key;
    client_max_body_size 2k;
    ssl_verify_client optional_no_ca;
    location / {
        proxy_pass https://127.0.0.1:8200;
        proxy_set_header Host $host;
        proxy_set_header X-SSL-CERT $ssl_client_escaped_cert;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto https;
        proxy_ssl_protocols TLSv1.3;
    }
}
```

Calling vault using curl via NGINX
```
{
  "errors": [
    "failed to parse the client certificate: x509: malformed certificate"
  ]
}
```

Headers that NGINX sends to vault
```
Host: vault.intra.hugorodrigues.xyz
X-SSL-CERT: -----BEGIN%20CERTIFICATE-----%0AMIICmzCCAk2gAwIBAgIUQmWTANF%2F2RW8W8itJ9fmf%2FfMMjgwBQYDK2VwMGAxCzAJ%0ABgNVBAYTAlBUMRcwFQYDVQQHEw5TYW1vcmEgQ29ycmVpYTESMBAGA1UECQwJU2Fu%0AdGFyw6ltMSQwIgYDVQQDExtSb2RyaWd1ZXMgSW50ZXJtZWRpYXRlIENBIDEwHhcN%0AMjMwNTEyMTUxNTE1WhcNMjQwNTExMTUxNTQ1WjCBrzELMAkGA1UEBhMCUFQxEjAQ%0ABgNVBAgMCVNhbnRhcsOpbTEXMBUGA1UEBxMOU2Ftb3JhIENvcnJlaWExHTAbBgNV%0ABAkTFFJ1YSBkYSBBbGVncmlhIDE0IDFFMREwDwYDVQQREwgyMTM1LTAyNjESMBAG%0AA1UEChMJUm9kcmlndWVzMS0wKwYDVQQDEyRzZWN1cml0eXB2MDEuaW50cmEuaHVn%0Ab3JvZHJpZ3Vlcy54eXowWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAAQQIPXeH1EA%0ACfMjST%2F%2FCAq23OemU%2F51yJKR9ljZrpMAMDyKmXmlW6yBzQoWQATalKohxpm4NrHZ%0AvvPQDQNk3zdno4GZMIGWMA4GA1UdDwEB%2FwQEAwIDqDATBgNVHSUEDDAKBggrBgEF%0ABQcDAjAdBgNVHQ4EFgQUjiYjBOFbWnOmvYa6TKqpuOM3Sq4wHwYDVR0jBBgwFoAU%0ACp6taklnERBo0%2BPgxl3vOYaLphYwLwYDVR0RBCgwJoIkc2VjdXJpdHlwdjAxLmlu%0AdHJhLmh1Z29yb2RyaWd1ZXMueHl6MAUGAytlcANBAL4vJC2mizp7PKpAyHr9d%2FDB%0AVt4ZLlYNMc%2FaEcANou0uX70FHNwgLScU%2FNmbtZzcoYS7G1F0Ch1stH33p2trhAg%3D%0A-----END%20CERTIFICATE-----%0A
X-Real-IP: 10.1.10.215
X-Forwarded-For: 10.1.10.215
X-Forwarded-Proto: https
Connection: close
user-agent: curl/8.1.0
accept: */*
```

JasonN3 commented:
@hmrodrigues, try replacing $ssl_client_escaped_cert with $ssl_client_raw_cert. In your current config, Nginx is modifying the certificate, which is why it's not able to be parsed.

hmrodrigues commented:
> @hmrodrigues, try replacing $ssl_client_escaped_cert with $ssl_client_raw_cert. In your current config, Nginx is modifying the certificate, which is why it's not able to be parsed.

If I change to raw I receive a 400 Bad Request

Headers sent by NGINX
```
Host: vault.intra.hugorodrigues.xyz
X-SSL-CERT: -----BEGIN CERTIFICATE-----

MIICmzCCAk2gAwIBAgIUQmWTANF/2RW8W8itJ9fmf/fMMjgwBQYDK2VwMGAxCzAJ
BgNVBAYTAlBUMRcwFQYDVQQHEw5TYW1vcmEgQ29ycmVpYTESMBAGA1UECQwJU2Fu
dGFyw6ltMSQwIgYDVQQDExtSb2RyaWd1ZXMgSW50ZXJtZWRpYXRlIENBIDEwHhcN
MjMwNTEyMTUxNTE1WhcNMjQwNTExMTUxNTQ1WjCBrzELMAkGA1UEBhMCUFQxEjAQ
BgNVBAgMCVNhbnRhcsOpbTEXMBUGA1UEBxMOU2Ftb3JhIENvcnJlaWExHTAbBgNV
BAkTFFJ1YSBkYSBBbGVncmlhIDE0IDFFMREwDwYDVQQREwgyMTM1LTAyNjESMBAG
A1UEChMJUm9kcmlndWVzMS0wKwYDVQQDEyRzZWN1cml0eXB2MDEuaW50cmEuaHVn
b3JvZHJpZ3Vlcy54eXowWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAAQQIPXeH1EA
CfMjST//CAq23OemU/51yJKR9ljZrpMAMDyKmXmlW6yBzQoWQATalKohxpm4NrHZ
vvPQDQNk3zdno4GZMIGWMA4GA1UdDwEB/wQEAwIDqDATBgNVHSUEDDAKBggrBgEF
BQcDAjAdBgNVHQ4EFgQUjiYjBOFbWnOmvYa6TKqpuOM3Sq4wHwYDVR0jBBgwFoAU
Cp6taklnERBo0+Pgxl3vOYaLphYwLwYDVR0RBCgwJoIkc2VjdXJpdHlwdjAxLmlu
dHJhLmh1Z29yb2RyaWd1ZXMueHl6MAUGAytlcANBAL4vJC2mizp7PKpAyHr9d/DB
Vt4ZLlYNMc/aEcANou0uX70FHNwgLScU/NmbtZzcoYS7G1F0Ch1stH33p2trhAg=
-----END CERTIFICATE-----
```

For what I can figure, every header after X-SSL-CERT isn't sent by NGINX. If I put `proxy_set_header X-SSL-CERT` at the end, the `Connection`, `user-agent` and `accept` are missing as well

```
Host: vault.intra.hugorodrigues.xyz
X-Real-IP: 10.1.10.215
X-Forwarded-For: 10.1.10.215
X-Forwarded-Proto: https
X-SSL-CERT: -----BEGIN CERTIFICATE-----

MIICmzCCAk2gAwIBAgIUQmWTANF/2RW8W8itJ9fmf/fMMjgwBQYDK2VwMGAxCzAJ
BgNVBAYTAlBUMRcwFQYDVQQHEw5TYW1vcmEgQ29ycmVpYTESMBAGA1UECQwJU2Fu
dGFyw6ltMSQwIgYDVQQDExtSb2RyaWd1ZXMgSW50ZXJtZWRpYXRlIENBIDEwHhcN
MjMwNTEyMTUxNTE1WhcNMjQwNTExMTUxNTQ1WjCBrzELMAkGA1UEBhMCUFQxEjAQ
BgNVBAgMCVNhbnRhcsOpbTEXMBUGA1UEBxMOU2Ftb3JhIENvcnJlaWExHTAbBgNV
BAkTFFJ1YSBkYSBBbGVncmlhIDE0IDFFMREwDwYDVQQREwgyMTM1LTAyNjESMBAG
A1UEChMJUm9kcmlndWVzMS0wKwYDVQQDEyRzZWN1cml0eXB2MDEuaW50cmEuaHVn
b3JvZHJpZ3Vlcy54eXowWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAAQQIPXeH1EA
CfMjST//CAq23OemU/51yJKR9ljZrpMAMDyKmXmlW6yBzQoWQATalKohxpm4NrHZ
vvPQDQNk3zdno4GZMIGWMA4GA1UdDwEB/wQEAwIDqDATBgNVHSUEDDAKBggrBgEF
BQcDAjAdBgNVHQ4EFgQUjiYjBOFbWnOmvYa6TKqpuOM3Sq4wHwYDVR0jBBgwFoAU
Cp6taklnERBo0+Pgxl3vOYaLphYwLwYDVR0RBCgwJoIkc2VjdXJpdHlwdjAxLmlu
dHJhLmh1Z29yb2RyaWd1ZXMueHl6MAUGAytlcANBAL4vJC2mizp7PKpAyHr9d/DB
Vt4ZLlYNMc/aEcANou0uX70FHNwgLScU/NmbtZzcoYS7G1F0Ch1stH33p2trhAg=
-----END CERTIFICATE-----
```

cipherboy commented:
Hmm, we might want to enable base64url decoding of headers then. 

Just for transparency, I didn't get as much time this release as I'd like to look at this, so sadly I think this will have to slip to early in 1.15 cycle.

hmrodrigues commented:
Passing both the raw header and the result of url.QueryUnescape to base64.URLEncoding.DecodeString also fails with malformed certificate

hmrodrigues commented:
yup now it fails to `failed to decode the client certificate: illegal base64 data at input byte 0` because nginx doesn't base64 encode the certificate.

se-chris-thach commented:
@JasonN3 Which version of Traefik are you using? It looks like it matters.

It appears that how Traefik encodes the header was changed in this commit https://github.com/traefik/traefik/commit/2c550c284d19b02e2ff4c9c9e066cf548349a033, affecting Traefik v2.9.2 and newer. 

I forked your branch `JasonN3:clientcert` and made some changes to adhere to this new encoding scheme and it works with Traefik v2.10.1 (latest at the time of writing this). See my changes [here](https://github.com/se-chris-thach/hashicorp-vault/commit/fa7d1dba9155b170bf974b471f3ece456c36314e).

The difference in Traefik v2.9.2 or newer is that the certificate PEM is no longer URL-escaped and the new lines and BEGIN-END statements are removed.

If someone has a better way of decoding this string, I'm totally open.

hmrodrigues commented:
> If someone has a better way of decoding this string, I'm totally open. It also kind of seems that how the certificate is encoded in this header is proxy dependent (Traefik, nginx, etc) based on all the comments.

Maybe add a second configuration to specify the decode method based on the used proxy?

hmrodrigues commented:
Also regarding nginx, just escaping the certificate also fails to parse it. Can it be related to being sent as PEM by nginx? [ParseCertificate](https://pkg.go.dev/crypto/x509#ParseCertificate) says ASN.1 DER

JasonN3 commented:
Rebased the code so it's up to date and added in an option `x_forwarded_for_client_cert_header_decoders`. This new option is a comma delimetered list of what actions need to be take in order to get the certificate to DER format. I tested with both Traefik and Nginx. For the latest version of Traefik, you just need to set it to `BASE64`. For Nginx, you need to set it to `URL,DER`. This will give flexibility for other proxies that may require different decode processes while not requiring Vault to figure out how to decode it every time there's a request.

I still need to update the tests, but this at least works.

JasonN3 commented:
Let me know if I should just recreate this PR since that rebase caused GitHub to flag it as a massive change even though it's the same 4 files that were changed.

cipherboy commented:
@JasonN3 Hmm, that's weird that it didn't behave nicely. Do you want me to try a rebase on my end and re-push up? I might not get to it until Tuesday though (and don't want to clobber any of your changes), but I think sometime next month I should have cycles to look at this and get it merged for 1.15. Let me know!

JasonN3 commented:
@cipherboy If you want to try, feel free. It is weird that it did that. I just added this repo as a remote labeled upstream and then ran `git rebase upstream/main`. If you clobber anything, I still have the code on my machine. I can also fix up anything that breaks.

cipherboy commented:
@JasonN3 Interesting, it looked like your top commit was:

```
commit dfdba2da40b64d5fb691d0a0148c70ffbba9d692
Merge: fe19996ed5 4761980179
Author: Jason N <33561705+JasonN3@users.noreply.github.com>
Date:   Fri May 26 11:08:58 2023 -0500

    Merge branch 'clientcert' of https://github.com/JasonN3/hashicorp-vault into clientcert
```

So perhaps there was a local merge of your locally rebased changes with the remote (not-updated) version of `clientcert`?

When I tried a rebase, I ended up getting a bunch of weird conflicts, so I ended up aborting it and stuck with a simple `git merge upstream/main` and that seemed to appease GH.

JasonN3 commented:
@cipherboy There were a few merge conflicts when I rebased. I just selected which code was correct in VSCode, but I guess it did something out of order to cause that issue or I had something wrong in my cloned repo. Thank you for fixing it. It looks a lot better without thousands of files changed.

JasonN3 commented:
I looked through the current resulting code. Everything looks to still be there and still builds and works. The section added to `website/content/docs/configuration/listener/tcp.mdx` got duplicated, but I'll fix that when I fix the tests

se-chris-thach commented:
Any update on when this might be included?

cwchristerw commented:
I'm looking forward to this and its good that there is support for using custom header :smile: I would like to know something, I have seen multiple guides like...

```
proxy_set_header X-SSL-CERT $ssl_client_escaped_cert;
```

https://nginx.org/en/docs/http/ngx_http_ssl_module.html#ssl_client_escaped_cert

cwchristerw commented on website/content/docs/configuration/listener/tcp.mdx:
```diff
@@ -180,6 +180,26 @@ default value in the `"/sys/config/ui"` [API endpoint](/vault/api-docs/system/co
   connecting client's IP, for example `3.4.5.6`. Note this requires the load balancer 
   to send the connecting client's IP in the `X-Forwarded-For` header.
 
+- `x_forwarded_for_client_cert_header` `(string: "")` –
```
I would use x_forwarded_for_tls_client_cert_header.

JasonN3 replied:
That name was picked to match the naming of the rest of the variables: `tls_require_and_verify_client_cert` and `tls_disable_client_certs`. For both of those variables, it is just `client_cert`.

cwchristerw commented on website/content/docs/configuration/listener/tcp.mdx:
```diff
@@ -180,6 +180,26 @@ default value in the `"/sys/config/ui"` [API endpoint](/vault/api-docs/system/co
+  This is required if you use the [TLS Certificates Auth Method](/docs/auth/cert) and your 
+  vault server is behind a reverse proxy.
+
+- `x_forwarded_for_client_cert_header_decoders` `(string: "")` –
```
I would use x_forwarded_for_tls_client_cert_header_decoders

JasonN3 replied:
That name was picked to match the naming of the rest of the variables: `tls_require_and_verify_client_cert` and `tls_disable_client_certs`. For both of those variables, it is just `client_cert`.

cwchristerw commented on http/forwarded_for_test.go:
```diff
@@ -255,4 +257,80 @@ func TestHandler_XForwardedFor(t *testing.T) {
+		req := client.NewRequest("GET", "/")
+		req.Headers = make(http.Header)
+		req.Headers.Set("x-forwarded-for", "5.6.7.8")
+		testcertificate := `MIIDtTCCAp2gAwIBAgIUf%2BjhKTFBnqSs34II0WS1L4QsbbAwDQYJKoZIhvcNAQEL%0ABQAwFjEUMBIGA1UEAxMLZXhhbXBsZS5jb20wHhcNMTYwMjI5MDIyNzQxWhcNMjUw%0AMTA1MTAyODExWjAbMRkwFwYDVQQDExBjZXJ0LmV4YW1wbGUuY29tMIIBIjANBgkq%0AhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAsZx0Svr82YJpFpIy4fJNW5fKA6B8mhxS%0ATRAVnygAftetT8puHflY0ss7Y6X2OXjsU0PRn%2B1PswtivhKi%2BeLtgWkUF9cFYFGn%0ASgMld6ZWRhNheZhA6ZfQmeM%2FBF2pa5HK2S ...
```
Would this work also
```
testcertificate := `-----BEGIN%20CERTIFICATE-----%0AMIIESDCCAzCgAwIBAgIUWWQIzsdICMRWpzb2yPYRJZVHaYgwDQYJKoZIhvcNAQEL%0ABQAwDzENMAsGA1UEAxMEVGVzdDAeFw0yMzA4MDIwNDE3NTZaFw0yMzA4MDIwNTE5%0AMTZaMBsxGTAXBgNVBAMTEG5vZGUxLmRvbWFpbi5jb20wggIiMA0GCSqGSIb3DQEB%0AAQUAA4ICDwAwggIKAoICAQCswKCIqCoKTvAySZnAnPPV8iGoS59xJnyYmJ4TChK0%0A1%2BhB8htli6L34Kh6aZ52YRMCXpKpXJEDmSjK3DtdivqOLYFoHPSe%2FBgs64NMT1Vp%0ApoWirh4oVCMb6ojGKHpYMczH3BzP5xAUfNPGWFqK5%2BI7YVQv7oR7gy8yR28bJz%2Bm%0A3Qpt5uyzCol5cC2HOepfbF6x44a%2BjLSVdNUbQxtL0n1RcNQgwur7o4c6F59gc1Bc%0Ao%2FTfJSPsWZ62ejaYBOVnV9SxvjNsw0TirxOKJ38tyHba1pTSISeaG57DD6LDku%2Bl%0Ailj3dPb6Fq8KudrgUVz9dfFeQHDhH%2F7q9fz0aaFlVbnYY6y7yIhW%2B1DJQLXTSacX%0AMq2OVwUYjlmx7NUS8bCVqwDiB4zUjlrAjE%2BxKtQ0s4B4lRNLsSoXKBT0F7N6sBBf%0ANP4BJSq6Zt5E3qMvPTPgWRAfT99m0tI13DfMWhPnAvu7kmiFsAMUzmD4B%2BUlb7ak%0A6qeyRglas9CXuVYfw7Y3s16zI7a8GSm%2BJfHxvhX70OsyTSh3CUP9WBCAmzC%2FY%2F8x%0A8jiCj87DFTXZkClCbPTH2tWukNeKg2eY0f8eL7zu2wKeBskKjakBb6SHBDkvGMah%0AyS%2F76TFpO1W8KABoGl%2Bu%2B%2BOOrz2eeQbQYTbRyJHPStOkOY6V%2F%2BA0TvqNlzWu12tD%0AnQIDAQABo4GPMIGMMA4GA1UdDwEB%2FwQEAwIDqDAdBgNVHSUEFjAUBggrBgEFBQcD%0AAQYIKwYBBQUHAwIwHQYDVR0OBBYEFIk038Eim%2Bvy%2BDPpobgdCUygwDcdMB8GA1Ud%0AIwQYMBaAFBved0eZwAgo2x1d8vOoVw3dLdDwMBsGA1UdEQQUMBKCEG5vZGUxLmRv%0AbWFpbi5jb20wDQYJKoZIhvcNAQELBQADggEBAJ1Ci7iU9bCApACMP73idO7zxCf7%0AGEmTsrRU5PgbOfJ3WMf0YBu6ynWPovGg2rh4N4KUs9cO0YJCmuerrZUIIQjPF%2FSa%0A42NWoAO35yogcXX6YS3ujz8eYfplh9rrQDQvDopJCpX6YeEY5oM2uEslCickiRI%2B%0AlhnedZ9tOdsJiAfheEeyEI%2BaQiZoXYYopMtmPHbcQ9jPRf54LeNlw%2FgX6o4BJV83%0AQhHcGh%2FrRYwjPE%2Farj2pUqY01b0z7zRAvn8g6Ju5qXzw6D3t507imFwEbE9WMFFe%0A24ZJVFCPjSN842GmrR56Dh3gR7TKXxstaNoSA0tmul9unu6tw8T%2Bx%2FtMzXQ%3D%0A-----END%20CERTIFICATE-----%0A`
```
with `-----BEGIN%20CERTIFICATE-----%0A` and `-----END%20CERTIFICATE-----%0A`

JasonN3 replied:
It will work with NGINX. See https://github.com/hashicorp/vault/pull/17272#issuecomment-1564623194

Cajga commented:
We are also looking forward to get this merged and would assign some developer time if needed to get this rolled.

@JasonN3 nice work and thanks for keeping it up!

se-chris-thach reviewed (approved):
LGTM

Gunni commented:
Please merge?

JasonN3 commented:
> Finally getting back around to reviewing this, sorry for the delay. Could you make one final tweak and fix the spelling around delimited in the documentation?

I just want to make sure I'm fixing what you're seeing. Is this the line you want fixed?
```
Comma delimeters list that specifies the decoders that will be used to decode the client certificate.
```
Do you just want it to say `Comma delimetered` instead of `Comma delimeters` or is there something else I'm overlooking?

JasonN3 commented:
@sgmiller I think I fixed the typo you wanted. Let me know if you meant something else or if there's anything else you want changed

cwchristerw commented:
I would like to make sure that this would be backported 1.14.x?

sgmiller commented:
This would be considered a new feature, and would not be backported.  We're just taking one more pass on the security of this, thanks for the change @JasonN3 .

sgmiller reviewed (approved):
Apologize to everyone interested in this feature for the time it's taken.  Approved.

JasonN3 commented:
@sgmiller, do you know what next steps are left? It's been a couple of weeks since your approval.

VioletHynes commented:
I'm going to hit update on the branch as there was previously an issue with our `build / completed-successfully` GHA that prevented it passing on community submitted PRs, as that seemed to be an issue blocking this PR, and that might have been an issue getting this merged. It has since been fixed, and the update should have everything passing as expected (except Vercel, unfortunately our hands are tied there).

I'll leave it up to @sgmiller as to when to merge this though as I've not been as close to this issue.

sgmiller commented:
@JasonN3 Do you want the honor of merging?  If not I'll merge it in a couple of days.

JasonN3 commented:
@sgmiller, do you know what next steps are left? It's been a couple of weeks since your approval.

> @JasonN3 Do you want the honor of merging? If not I'll merge it in a couple of days.

The only merge I can do is merge the updates to main into my fork. I can't merge my fork into this repo.

sgmiller commented:
> @sgmiller, do you know what next steps are left? It's been a couple of weeks since your approval.
> 
> > @JasonN3 Do you want the honor of merging? If not I'll merge it in a couple of days.
> 
> The only merge I can do is merge the updates to main into my fork. I can't merge my fork into this repo.

Doh, right.  Okay, I'll merge.
